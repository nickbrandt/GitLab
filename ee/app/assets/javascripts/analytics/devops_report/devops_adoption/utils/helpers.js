import { isToday } from '~/lib/utils/datetime_utility';

/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, timestamp and check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, timestamp, openModal }) => {
  if (openModal) {
    return false;
  } else if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );
  const dataNotRefreshedToday = !isToday(new Date(timestamp));

  return anyPendingEnabledNamespaces || dataNotRefreshedToday;
};
