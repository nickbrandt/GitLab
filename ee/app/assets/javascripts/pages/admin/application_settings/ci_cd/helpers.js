/**
 * Filters [gitlabCiYmls] based on a given [searchTerm].
 * Yml catagories with no items after filtering are not included in the returned object.
 * @param {Object} gitlabCiYmls - { <categoryName>: [{ name, id }] }
 * @param {String} searchTerm
 * @returns {Object}
 */
export function filterGitlabCiYmls(gitlabCiYmls, searchTerm) {
  return Object.keys(gitlabCiYmls).reduce((filteredYmls, category) => {
    const categoryYmls = gitlabCiYmls[category].filter((yml) =>
      yml.name.toLowerCase().startsWith(searchTerm),
    );

    if (categoryYmls.length > 0) {
      Object.assign(filteredYmls, {
        [category]: categoryYmls,
      });
    }

    return filteredYmls;
  }, {});
}
