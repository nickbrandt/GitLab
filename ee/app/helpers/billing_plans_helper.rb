# frozen_string_literal: true

module BillingPlansHelper
  include Gitlab::Utils::StrongMemoize

  def subscription_plan_info(plans_data, current_plan_code)
    current_plan = plans_data.find { |plan| plan.code == current_plan_code && plan.current_subscription_plan? }
    current_plan || plans_data.find { |plan| plan.code == current_plan_code }
  end

  def number_to_plan_currency(value)
    number_to_currency(value, unit: '$', strip_insignificant_zeros: true, format: "%u%n")
  end

  def upgrade_offer_type(namespace, plan)
    return :no_offer if namespace.actual_plan_name != Plan::BRONZE || !offer_from_previous_tier?(namespace.id, plan.id)

    upgrade_for_free?(namespace.id) ? :upgrade_for_free : :upgrade_for_offer
  end

  def has_upgrade?(upgrade_offer)
    upgrade_offer == :upgrade_for_free || upgrade_offer == :upgrade_for_offer
  end

  def show_contact_sales_button?(purchase_link_action, upgrade_offer)
    return false unless purchase_link_action == 'upgrade'

    upgrade_offer == :upgrade_for_offer ||
      (experiment_enabled?(:contact_sales_btn_in_app) && upgrade_offer == :no_offer)
  end

  def show_upgrade_button?(purchase_link_action, upgrade_offer)
    purchase_link_action == 'upgrade' &&
      (upgrade_offer == :no_offer || upgrade_offer == :upgrade_for_free)
  end

  def subscription_plan_data_attributes(namespace, plan)
    return {} unless namespace

    {
      namespace_id: namespace.id,
      namespace_name: namespace.name,
      add_seats_href: add_seats_url(namespace),
      plan_upgrade_href: plan_upgrade_url(namespace, plan),
      plan_renew_href: plan_renew_url(namespace),
      customer_portal_url: "#{EE::SUBSCRIPTIONS_URL}/subscriptions",
      billable_seats_href: billable_seats_href(namespace),
      plan_name: plan&.name,
      free_personal_namespace: namespace.free_personal?.to_s
    }
  end

  def use_new_purchase_flow?(namespace)
    # new flow requires the user to already have a last name.
    # This can be removed once https://gitlab.com/gitlab-org/gitlab/-/issues/298715 is complete.
    return false unless current_user.last_name.present?

    namespace.group? && (namespace.actual_plan_name == Plan::FREE || namespace.trial_active?)
  end

  def experiment_tracking_data_for_button_click(button_label)
    return {} unless Gitlab::Experimentation.active?(:contact_sales_btn_in_app)

    {
      track: {
        event: 'click_button',
        label: button_label,
        property: experiment_tracking_category_and_group(:contact_sales_btn_in_app)
      }
    }
  end

  def plan_feature_list(plan)
    return [] unless plan.features

    plan.features.sort_by! { |feature| feature.highlight ? 0 : 1 }
  end

  def plan_purchase_or_upgrade_url(group, plan)
    if group.upgradable?
      plan_upgrade_url(group, plan)
    else
      plan_purchase_url(group, plan)
    end
  end

  def show_plans?(namespace)
    return false if namespace.free_personal?

    namespace.trial_active? || !(namespace.gold_plan? || namespace.ultimate_plan?)
  end

  def show_trial_banner?(namespace)
    return false unless params[:trial]

    root = namespace.has_parent? ? namespace.root_ancestor : namespace
    root.trial_active?
  end

  def namespace_for_user?(namespace)
    namespace == current_user.namespace
  end

  def seats_data_last_update_info
    last_enqueue_time = UpdateMaxSeatsUsedForGitlabComSubscriptionsWorker.last_enqueue_time&.utc
    return _("Seats usage data as of %{last_enqueue_time} (Updated daily)" % { last_enqueue_time: last_enqueue_time }) if last_enqueue_time

    _('Seats usage data is updated every day at 12:00pm UTC')
  end

  def upgrade_button_css_classes(namespace, plan, is_current_plan)
    css_classes = %w[btn btn-success gl-button]

    css_classes << 'disabled' if is_current_plan && !namespace.trial_active?
    css_classes << 'invisible' if ::Feature.enabled?(:hide_deprecated_billing_plans) && plan.deprecated?

    css_classes.join(' ')
  end

  def billing_available_plans(plans_data, current_plan)
    return plans_data unless ::Feature.enabled?(:hide_deprecated_billing_plans)

    plans_data.reject do |plan_data|
      if plan_data.code == current_plan&.code
        plan_data.deprecated? && plan_data.hide_deprecated_card?
      else
        plan_data.deprecated?
      end
    end
  end

  def show_start_free_trial_messages?(namespace)
    !namespace.free_personal? && namespace.eligible_for_trial?
  end

  def plan_purchase_url(group, plan)
    if use_new_purchase_flow?(group)
      new_subscriptions_path(plan_id: plan.id, namespace_id: group.id, source: params[:source])
    else
      "#{plan.purchase_link.href}&gl_namespace_id=#{group.id}"
    end
  end

  private

  def add_seats_url(group)
    return unless group

    "#{EE::SUBSCRIPTIONS_URL}/gitlab/namespaces/#{group.id}/extra_seats"
  end

  def plan_upgrade_url(group, plan)
    return unless group && plan&.id

    "#{EE::SUBSCRIPTIONS_URL}/gitlab/namespaces/#{group.id}/upgrade/#{plan.id}"
  end

  def plan_renew_url(group)
    return unless group

    "#{EE::SUBSCRIPTIONS_URL}/gitlab/namespaces/#{group.id}/renew"
  end

  def billable_seats_href(namespace)
    return unless namespace.group?

    group_seat_usage_path(namespace)
  end

  def offer_from_previous_tier?(namespace_id, plan_id)
    upgrade_plan_id = upgrade_plan_data(namespace_id)[:upgrade_plan_id]

    return false unless upgrade_plan_id

    upgrade_plan_id == plan_id
  end

  def upgrade_for_free?(namespace_id)
    !!upgrade_plan_data(namespace_id)[:upgrade_for_free]
  end

  def upgrade_plan_data(namespace_id)
    strong_memoize(:upgrade_plan_data) do
      GitlabSubscriptions::PlanUpgradeService
        .new(namespace_id: namespace_id)
        .execute
    end
  end
end
