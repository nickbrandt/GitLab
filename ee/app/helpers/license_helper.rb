# frozen_string_literal: true

module LicenseHelper
  include ActionView::Helpers::AssetTagHelper
  include ActionView::Helpers::UrlHelper

  delegate :new_admin_license_path, to: 'Gitlab::Routing.url_helpers'

  def seats_calculation_message(license)
    return unless license.exclude_guests_from_active_count?

    s_("Users with a Guest role or those who don't belong to a Project or Group will not use a seat from your license.")
  end

  def current_license_title
    License.current&.plan&.titleize || 'Core'
  end

  def has_active_license?
    License.current.present?
  end

  def new_trial_url
    return_to_url = CGI.escape(Gitlab.config.gitlab.url)
    uri = URI.parse(::EE::SUBSCRIPTIONS_URL)
    uri.path = '/trials/new'
    uri.query = "return_to=#{return_to_url}&id=#{Base64.strict_encode64(current_user.email)}"
    uri.to_s
  end

  def show_promotions?(selected_user = current_user)
    return false unless selected_user

    if Gitlab::CurrentSettings.current_application_settings
      .should_check_namespace_plan?
      true
    else
      license = License.current
      license.nil? || license.expired?
    end
  end

  def show_advanced_search_promotion?
    !Gitlab::CurrentSettings.should_check_namespace_plan? && show_promotions? && show_callout?('promote_advanced_search_dismissed') && !License.feature_available?(:elastic_search)
  end

  def licensed_users(license)
    if license.restricted?(:active_user_count)
      number_with_delimiter(license.restrictions[:active_user_count])
    else
      _('Unlimited')
    end
  end

  def cloud_license_view_data
    {
      buy_subscription_path: ::EE::SUBSCRIPTIONS_PLANS_URL,
      customers_portal_url: ::EE::SUBSCRIPTIONS_MANAGE_URL,
      free_trial_path: new_trial_url,
      has_active_license: (has_active_license? ? 'true' : 'false'),
      license_upload_path: new_admin_license_path,
      license_remove_path: admin_license_path,
      subscription_sync_path: sync_seat_link_admin_license_path,
      congratulation_svg_path: image_path('illustrations/illustration-congratulation-purchase.svg'),
      subscription_activation_banner_callout_name: ::EE::UserCalloutsHelper::CL_SUBSCRIPTION_ACTIVATION
    }
  end

  extend self
end
