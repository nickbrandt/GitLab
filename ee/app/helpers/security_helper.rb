# frozen_string_literal: true

module SecurityHelper
  def instance_security_dashboard_data
    instance_security_dashboard = InstanceSecurityDashboard.new(current_user)

    {
      dashboard_documentation: help_page_path('user/application_security/security_dashboard/index', anchor: 'instance-security-dashboard'),
      no_vulnerabilities_svg_path: image_path('illustrations/issues.svg'),
      empty_dashboard_state_svg_path: image_path('illustrations/security-dashboard-empty-state.svg'),
      empty_state_svg_path: image_path('illustrations/operations-dashboard_empty.svg'),
      survey_request_svg_path: image_path('illustrations/security-dashboard_empty.svg'),
      project_add_endpoint: security_projects_path,
      project_list_endpoint: security_projects_path,
      instance_dashboard_settings_path: settings_security_dashboard_path,
      vulnerabilities_export_endpoint: expose_path(api_v4_security_vulnerability_exports_path),
      scanners: VulnerabilityScanners::ListService.new(instance_security_dashboard).execute.to_json,
      has_projects: instance_security_dashboard.projects.any?.to_s
    }
  end

  def security_dashboard_unavailable_view_data
    {
      empty_state_svg_path: image_path('illustrations/security-dashboard-empty-state.svg'),
      dashboard_documentation: help_page_path('user/application_security/security_dashboard/index'),
      is_unavailable: "true"
    }
  end

  def instance_security_settings_data
    {
      is_auditor: current_user.auditor?.to_s
    }
  end
end
