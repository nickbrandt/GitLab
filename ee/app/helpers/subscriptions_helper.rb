# frozen_string_literal: true

module SubscriptionsHelper
  include ::Gitlab::Utils::StrongMemoize

  def subscription_data(eligible_groups)
    {
      setup_for_company: (current_user.setup_for_company == true).to_s,
      full_name: current_user.name,
      available_plans: subscription_available_plans.to_json,
      plan_id: params[:plan_id],
      namespace_id: params[:namespace_id],
      new_user: new_user?.to_s,
      group_data: present_groups(eligible_groups).to_json,
      source: params[:source]
    }
  end

  def plan_title
    strong_memoize(:plan_title) do
      plan = subscription_available_plans.find { |plan| plan[:id] == params[:plan_id] }
      plan[:code].titleize if plan
    end
  end

  private

  def new_user?
    return false unless request.referer.present?

    URI.parse(request.referer).path == users_sign_up_welcome_path
  end

  def plans_data
    GitlabSubscriptions::FetchSubscriptionPlansService.new(plan: :free).execute
      .map(&:symbolize_keys)
      .reject { |plan_data| plan_data[:free] }
      .map { |plan_data| plan_data.slice(:id, :code, :price_per_year, :deprecated, :name, :hide_card) }
  end

  def subscription_available_plans
    return plans_data unless ::Feature.enabled?(:hide_deprecated_billing_plans)

    plans_data.reject { |plan_data| plan_data[:deprecated] || plan_data[:hide_card] }
  end

  def present_groups(groups)
    groups.map do |namespace|
      {
        id: namespace.id,
        name: namespace.name,
        users: namespace.member_count,
        guests: namespace.guest_count
      }
    end
  end
end
