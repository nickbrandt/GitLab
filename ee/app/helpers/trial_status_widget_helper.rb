# frozen_string_literal: true

# NOTE: The patterns first introduced in this helper for doing trial-related
# callouts are mimicked by the PaidFeatureCalloutHelper. A third reuse of these
# patterns (especially as these experiments finish & become permanent parts of
# the codebase) could trigger the need to extract these patterns into a single,
# reusable, sharable helper.
module TrialStatusWidgetHelper
  D14_CALLOUT_RANGE = (7..14).freeze # between 14 & 7 days remaining
  D3_CALLOUT_RANGE = (0..3).freeze # between 3 & 0 days remaining

  # NOTE: We are okay hard-coding the production value for the Ulitmate 1-year
  # SaaS plan ID while this is all part of an active experiment. If & when the
  # experiment is deemed a success, part of the clean-up effort will be to
  # pull the value directly from the CustomersDot API. Value taken from
  # https://gitlab.com/gitlab-org/customers-gitlab-com/blob/7177f13c478ef623b779d6635c4a58ee650b7884/config/application.yml#L207
  ZUORA_ULTIMATE_PLAN_ID = '2c92a0ff76f0d5250176f2f8c86f305a'

  def trial_status_popover_data_attrs(group)
    base_attrs = trial_status_common_data_attrs(group)
    base_attrs.merge(
      group_name: group.name,
      purchase_href: ultimate_subscription_path_for_group(group),
      start_initially_shown: in_forced_popover_experiment?(group) && force_popover_to_be_shown?(group.trial_days_remaining),
      target_id: base_attrs[:container_id],
      trial_end_date: group.trial_ends_on
    )
  end

  def trial_status_widget_data_attrs(group)
    trial_status_common_data_attrs(group).merge(
      days_remaining: group.trial_days_remaining,
      nav_icon_image_path: image_path('illustrations/golden_tanuki.svg'),
      percentage_complete: group.trial_percentage_complete
    )
  end

  def show_trial_status_widget?(group)
    billing_plans_and_trials_available? && eligible_for_trial_upgrade_callout?(group)
  end

  private

  def billing_plans_and_trials_available?
    ::Gitlab::CurrentSettings.should_check_namespace_plan?
  end

  def eligible_for_trial_upgrade_callout?(group)
    group.trial_active? && can?(current_user, :admin_namespace, group)
  end

  def force_popover_to_be_shown?(days_remaining)
    D14_CALLOUT_RANGE.cover?(days_remaining) || D3_CALLOUT_RANGE.cover?(days_remaining)
  end

  def trial_status_common_data_attrs(group)
    {
      container_id: 'trial-status-sidebar-widget',
      plan_name: group.gitlab_subscription&.plan_title,
      plans_href: group_billings_path(group)
    }
  end

  def ultimate_subscription_path_for_group(group)
    new_subscriptions_path(namespace_id: group.id, plan_id: ZUORA_ULTIMATE_PLAN_ID)
  end

  def in_forced_popover_experiment?(group)
    experiment(:forcibly_show_trial_status_popover, group: group).variant.group == :experiment
  end
end
