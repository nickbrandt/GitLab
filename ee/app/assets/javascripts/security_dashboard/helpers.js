import isPlainObject from 'lodash/isPlainObject';
import { REPORT_TYPES, SEVERITY_LEVELS } from 'ee/security_dashboard/store/constants';
import { BASE_FILTERS } from 'ee/security_dashboard/store/modules/filters/constants';
import convertReportType from 'ee/vue_shared/security_reports/store/utils/convert_report_type';
import { VULNERABILITY_STATES } from 'ee/vulnerabilities/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { s__, __ } from '~/locale';
import { DEFAULT_SCANNER } from './constants';

const parseOptions = (obj) =>
  Object.entries(obj).map(([id, name]) => ({ id: id.toUpperCase(), name }));

export const mapProjects = (projects) =>
  projects.map((p) => ({ id: p.id.split('/').pop(), name: p.name }));

const stateOptions = parseOptions(VULNERABILITY_STATES);
const defaultStateOptions = stateOptions.filter((x) => ['DETECTED', 'CONFIRMED'].includes(x.id));

export const stateFilter = {
  name: s__('SecurityReports|Status'),
  id: 'state',
  options: stateOptions,
  allOption: BASE_FILTERS.state,
  defaultOptions: defaultStateOptions,
};

export const severityFilter = {
  name: s__('SecurityReports|Severity'),
  id: 'severity',
  options: parseOptions(SEVERITY_LEVELS),
  allOption: BASE_FILTERS.severity,
  defaultOptions: [],
};

export const createScannerOption = (vendor, reportType) => {
  const type = reportType.toUpperCase();

  return {
    id: `${vendor}.${type}`,
    reportType: reportType.toUpperCase(),
    name: convertReportType(reportType),
    scannerIds: [],
  };
};

// This is used on the pipeline security tab, group-level report, and instance-level report. It's
// used by the scanner filter that shows a flat list of scan types (DAST, SAST, etc) with no vendor
// grouping.
export const simpleScannerFilter = {
  name: s__('SecurityReports|Scanner'),
  id: 'reportType',
  options: parseOptions(REPORT_TYPES),
  allOption: BASE_FILTERS.report_type,
  defaultOptions: [],
};

// This is used on the project-level report. It's used by the scanner filter that shows a list of
// scan types (DAST, SAST, etc) that's grouped by vendor.
export const vendorScannerFilter = {
  name: s__('SecurityReports|Scanner'),
  id: 'scanner',
  options: Object.keys(REPORT_TYPES).map((x) => createScannerOption(DEFAULT_SCANNER, x)),
  allOption: BASE_FILTERS.report_type,
  defaultOptions: [],
};

export const activityOptions = {
  NO_ACTIVITY: { id: 'NO_ACTIVITY', name: s__('SecurityReports|No activity') },
  WITH_ISSUES: { id: 'WITH_ISSUES', name: s__('SecurityReports|With issues') },
  NO_LONGER_DETECTED: { id: 'NO_LONGER_DETECTED', name: s__('SecurityReports|No longer detected') },
};

export const activityFilter = {
  name: s__('Reports|Activity'),
  id: 'activity',
  options: Object.values(activityOptions),
  allOption: BASE_FILTERS.activity,
  defaultOptions: [],
};

export const getProjectFilter = (projects) => {
  return {
    name: s__('SecurityReports|Project'),
    id: 'projectId',
    options: mapProjects(projects),
    allOption: BASE_FILTERS.project_id,
    defaultOptions: [],
  };
};

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

/**
 * We have disabled loading hasNextPage from GraphQL as it causes timeouts in database,
 * instead we have to calculate that value based on the existence of endCursor. When endCursor
 * is empty or has null value, that means that there is no next page to be loaded from GraphQL API.
 *
 * @param {Object} pageInfo
 * @returns {Object}
 */
export const preparePageInfo = (pageInfo) => {
  return { ...pageInfo, hasNextPage: Boolean(pageInfo?.endCursor) };
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

export default () => ({});
