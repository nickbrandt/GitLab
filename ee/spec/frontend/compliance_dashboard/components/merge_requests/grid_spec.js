import Approvers from 'ee/compliance_dashboard/components/merge_requests/approvers.vue';
import MergeRequestsGrid from 'ee/compliance_dashboard/components/merge_requests/grid.vue';
import Status from 'ee/compliance_dashboard/components/merge_requests/status.vue';
import BranchDetails from 'ee/compliance_dashboard/components/shared/branch_details.vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import TimeAgoTooltip from '~/vue_shared/components/time_ago_tooltip.vue';
import { createMergeRequests, mergedAt } from '../../mock_data';

describe('MergeRequestsGrid component', () => {
  let wrapper;

  const findMergeRequestDrawerToggles = () =>
    wrapper.findAllByTestId('merge-request-drawer-toggle');
  const findMergeRequests = () => wrapper.findAllByTestId('merge-request');
  const findMergeRequestLinks = () => wrapper.findAllByTestId('merge-request-link');
  const findTime = () => wrapper.findComponent(TimeAgoTooltip);
  const findStatuses = () => wrapper.findAllComponents(Status);
  const findApprovers = () => wrapper.findComponent(Approvers);
  const findBranchDetails = () => wrapper.findComponent(BranchDetails);

  const createComponent = (mergeRequests = {}, drawerEnabled = false) => {
    return shallowMountExtended(MergeRequestsGrid, {
      propsData: {
        mergeRequests,
        isLastPage: false,
        drawerEnabled,
      },
      stubs: {
        MergeRequest: {
          props: { mergeRequest: Object },
          template: `<div data-testid="merge-request"><a href="" data-testid="merge-request-link">{{ mergeRequest.title }}</a></div>`,
        },
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  // TODO: Remove the each with https://gitlab.com/gitlab-org/gitlab/-/issues/334682
  describe.each([true, false])('when drawer enabled is %s', (drawerEnabled) => {
    describe('when initialized', () => {
      beforeEach(() => {
        wrapper = createComponent(createMergeRequests({ count: 2, props: {} }), drawerEnabled);
      });

      it('matches the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('renders a list of merge requests', () => {
        expect(findMergeRequests()).toHaveLength(2);
      });

      it('renders the approvers list', () => {
        expect(findApprovers().exists()).toBe(true);
      });

      it('renders the "merged at" time', () => {
        expect(findTime().props('time')).toEqual(mergedAt());
      });
    });

    describe('statuses', () => {
      const mergeRequest = createMergeRequests({ count: 1 });

      beforeEach(() => {
        wrapper = createComponent(mergeRequest, drawerEnabled);
      });

      it('passes the correct props to the statuses', () => {
        findStatuses().wrappers.forEach((status) => {
          const { type, data } = status.props('status');

          switch (type) {
            case 'pipeline':
              expect(data).toEqual(mergeRequest[0].pipeline_status);
              break;

            case 'approval':
              expect(data).toEqual(mergeRequest[0].approval_status);
              break;

            default:
              throw new Error('Unknown status type');
          }
        });
      });
    });

    describe('branch details', () => {
      it('does not render if there are no branch details', () => {
        wrapper = createComponent(createMergeRequests({ count: 2, props: {} }), drawerEnabled);

        expect(findBranchDetails().exists()).toBe(false);
      });

      it('renders if there are branch details', () => {
        wrapper = createComponent(
          createMergeRequests({
            count: 2,
            props: { target_branch: 'main', source_branch: 'feature' },
          }),
          drawerEnabled,
        );

        expect(findBranchDetails().exists()).toBe(true);
      });
    });
  });

  describe('when the drawer is enabled', () => {
    const mergeRequests = createMergeRequests({ count: 2, props: {} });

    beforeEach(() => {
      const mergeRequest = createMergeRequests({ count: 1 });
      wrapper = createComponent(mergeRequest, true);
    });

    describe.each(['click', 'keypress.enter'])('when the %s event is triggered', (event) => {
      it('toggles the drawer when a merge request drawer toggle is the target', () => {
        findMergeRequestDrawerToggles().at(0).trigger(event);

        expect(wrapper.emitted('toggleDrawer')[0][0]).toStrictEqual(mergeRequests[0]);
      });

      it('does not toggle the drawer if an inner link is the target', () => {
        findMergeRequestLinks().at(0).trigger(event);

        expect(wrapper.emitted('toggleDrawer')).toBe(undefined);
      });
    });
  });
});
