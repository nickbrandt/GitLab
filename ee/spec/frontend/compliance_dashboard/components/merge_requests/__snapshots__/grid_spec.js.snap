// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeRequestsGrid component when drawer enabled is false when initialized matches the snapshot 1`] = `
<div>
  <div
    class="dashboard-grid gl-display-grid gl-grid-tpl-rows-auto"
  >
    <grid-column-heading-stub
      heading="Merge Request"
    />
     
    <grid-column-heading-stub
      class="gl-text-center"
      heading="Approval Status"
    />
     
    <grid-column-heading-stub
      class="gl-text-center"
      heading="Pipeline"
    />
     
    <grid-column-heading-stub
      class="gl-text-right"
      heading="Updates"
    />
     
    <div
      data-testid="merge-request"
    >
      <a
        data-testid="merge-request-link"
        href=""
      >
        Merge request 0
      </a>
    </div>
     
    <status-stub
      status="[object Object]"
    />
     
    <status-stub
      status="[object Object]"
    />
     
    <div
      class="gl-text-right gl-border-b-solid gl-border-b-1 gl-border-b-gray-100 gl-p-5 gl-relative"
    >
      <approvers-stub
        approvers=""
      />
       
      <!---->
       
      <time-ago-tooltip-stub
        class="gl-text-gray-500"
        cssclass=""
        time="2020-01-01T00:00:00.000Z"
        tooltipplacement="bottom"
      />
    </div>
    <div
      data-testid="merge-request"
    >
      <a
        data-testid="merge-request-link"
        href=""
      >
        Merge request 1
      </a>
    </div>
     
    <status-stub
      status="[object Object]"
    />
     
    <status-stub
      status="[object Object]"
    />
     
    <div
      class="gl-text-right gl-border-b-solid gl-border-b-1 gl-border-b-gray-100 gl-p-5 gl-relative"
    >
      <approvers-stub
        approvers=""
      />
       
      <!---->
       
      <time-ago-tooltip-stub
        class="gl-text-gray-500"
        cssclass=""
        time="2020-01-01T00:00:00.000Z"
        tooltipplacement="bottom"
      />
    </div>
  </div>
   
  <pagination-stub
    class="gl-mt-5"
  />
</div>
`;

exports[`MergeRequestsGrid component when drawer enabled is true when initialized matches the snapshot 1`] = `
<div>
  <div
    class="dashboard-grid gl-display-grid gl-grid-tpl-rows-auto"
  >
    <grid-column-heading-stub
      heading="Merge Request"
    />
     
    <grid-column-heading-stub
      class="gl-text-center"
      heading="Approval Status"
    />
     
    <grid-column-heading-stub
      class="gl-text-center"
      heading="Pipeline"
    />
     
    <grid-column-heading-stub
      class="gl-text-right"
      heading="Updates"
    />
     
    <div
      class="dashboard-merge-request dashboard-grid gl-display-grid gl-grid-tpl-rows-auto gl-hover-bg-blue-50 gl-hover-text-decoration-none gl-hover-cursor-pointer"
      data-testid="merge-request-drawer-toggle"
      tabindex="0"
    >
      <div
        data-testid="merge-request"
      >
        <a
          data-testid="merge-request-link"
          href=""
        >
          Merge request 0
        </a>
      </div>
       
      <status-stub
        status="[object Object]"
      />
       
      <status-stub
        status="[object Object]"
      />
       
      <div
        class="gl-text-right gl-border-b-solid gl-border-b-1 gl-border-b-gray-100 gl-p-5 gl-relative"
      >
        <approvers-stub
          approvers=""
        />
         
        <!---->
         
        <time-ago-tooltip-stub
          class="gl-text-gray-500"
          cssclass=""
          time="2020-01-01T00:00:00.000Z"
          tooltipplacement="bottom"
        />
      </div>
    </div>
    <div
      class="dashboard-merge-request dashboard-grid gl-display-grid gl-grid-tpl-rows-auto gl-hover-bg-blue-50 gl-hover-text-decoration-none gl-hover-cursor-pointer"
      data-testid="merge-request-drawer-toggle"
      tabindex="0"
    >
      <div
        data-testid="merge-request"
      >
        <a
          data-testid="merge-request-link"
          href=""
        >
          Merge request 1
        </a>
      </div>
       
      <status-stub
        status="[object Object]"
      />
       
      <status-stub
        status="[object Object]"
      />
       
      <div
        class="gl-text-right gl-border-b-solid gl-border-b-1 gl-border-b-gray-100 gl-p-5 gl-relative"
      >
        <approvers-stub
          approvers=""
        />
         
        <!---->
         
        <time-ago-tooltip-stub
          class="gl-text-gray-500"
          cssclass=""
          time="2020-01-01T00:00:00.000Z"
          tooltipplacement="bottom"
        />
      </div>
    </div>
  </div>
   
  <pagination-stub
    class="gl-mt-5"
  />
</div>
`;
